cpp/src/community/detail/mis_impl.cuh

```diff
@@ -117,22 +104,47 @@ rmm::device_uvector<vertex_t> compute_mis(
     thrust::copy(handle.get_thrust_policy(), ranks.begin(), ranks.end(), temporary_ranks.begin());
 
     // Select a random set of candidate vertices
-    // FIXME: use common utility function to select a subset of remaining vertices
-    // and for MG extension, select from disributed array remaining vertices
-    thrust::shuffle(
-      handle.get_thrust_policy(), remaining_vertices.begin(), remaining_vertices.end(), g);
 
-    vertex_t nr_candidates =
-      (remaining_vertices.size() < 1024)
-        ? remaining_vertices.size()
-        : std::min(static_cast<vertex_t>((0.50 + 0.25 * loop_counter) * remaining_vertices.size()),
-                   static_cast<vertex_t>(remaining_vertices.size()));
+    vertex_t nr_remaining_vertices_to_check = remaining_vertices.size();
```

Comment:
I think this is functional.  I wonder if this could be done more efficiently.  Perhaps a FIXME to review the performance here.

My thought on a more efficient implementation would be this:

1. Outside of the loop, on each GPU, randomly shuffle the original remaining_vertices array
2. Outside of the loop, do the `host_scalar_allreduce` call (or just ask the graph for the total number of vertices)
3. Inside the loop, decide on the next batch of 1024 elements to process.  Does it matter that this is truly random?  If not, you could just round-robin grab 1024 vertices to process.  If it does, you could determine which GPUs to select the next 1024 vertices from.
4. For GPU X, you now know (from step 3) how many elements (k) you are using from its remaining_vertices set, take the next k elements from the list to operate on.  Update a local pointer to keep track of where you are so the next iteration continues working down the randomly ordered list.
5. Decrement the global vertex counter by 1024 - you know that's how many vertices you need

I feel like this would be more efficient (reduce the number of comms calls, reduce the number of sorts and other function calls required by the `select_random_vertices` call).

But again, this appears functional, I'm inclined to use this approach and just add a FIXME to improve performance later.

Reply:
I guess here, we are selecting at least 1024, not just 1024. As the loop count grows, we may select all the remaining vertices.

And I agree with Chuck that "inclined to use this approach and just add a FIXME to improve performance later." if this turns out to be a performance bottleneck later.